CefSharp.BrowserSubprocess.Core/CefAppUnmanagedWrapper.cpp

```diff
@@ -91,64 +91,105 @@ namespace CefSharp
+                    //we need to do this here to be able to store the v8context
+                    if (success)
                     {
-                        CefRefPtr<CefV8Value> result;
-                        CefRefPtr<CefV8Exception> exception;
-                        auto success = context->Eval(script, result, exception);
-                        auto response = CefProcessMessage::Create(kEvaluateJavascriptResponse);
                         auto argList = response->GetArgumentList();
+                        SerializeV8Object(result, argList, 2, browserWrapper->CallbackRegistry);
+                    }
+                }
+                else
+                {
+                    //TODO handle error
+                }
+            }
+            else
+            {
+                auto jsCallbackId = GetInt64(argList, 1);
+                callbackId = GetInt64(argList, 2);
+                auto parameterList = argList->GetList(3);
+                CefV8ValueList params;
+                for (CefV8ValueList::size_type i = 0; i < parameterList->GetSize(); i++)
+                {
+                    params.push_back(DeserializeV8Object(parameterList, static_cast<int>(i)));
+                }
+
+                auto callbackRegistry = browserWrapper->CallbackRegistry;
+                auto callbackWrapper = callbackRegistry->FindWrapper(jsCallbackId);
+                auto context = callbackWrapper->GetContext();
+                auto value = callbackWrapper->GetValue();
+
+                {
+                    V8ContextScope scope(context);
 
-                        argList->SetBool(0, success);
-                        SetInt64(callbackId, argList, 1);
-                        if (success)
-                        {
-                            SerializeV8Object(result, argList, 2, browserWrapper->CallbackRegistry);
-                        }
-                        else
-                        {
-                            argList->SetString(2, exception->GetMessage());
-                        }
-
-                        if (response.get())
-                        {
-                            browser->SendProcessMessage(sourceProcessId, response);
-                        }
+                    result = value->ExecuteFunction(nullptr, params);
+                    success = result.get() != nullptr;
+                    response = CefProcessMessage::Create(kJavascriptCallbackResponse);
+                    //we need to do this here to be able to store the v8context
+                    if (success)
+                    {
+                        auto argList = response->GetArgumentList();
+                        SerializeV8Object(result, argList, 2, browserWrapper->CallbackRegistry);
                     }
-                    finally
+                    else
                     {
-                        context->Exit();
+                        exception = value->GetException();
```

Comment:
Rather than store the exception it might be better to store a `CefString`, then set that if an exception occurs. Cases when the matching frame wasn't found etc could also be dealt with.